### ISSUE TITLE
Commands: .NET Core CLI Implementation

### ISSUE THREAD

**Issue by bricelam** [aspnet/EntityFramework#3925]
DNX commands will be going away with the transition to [.NET Core CLI](https://github.com/dotnet/cli). We'll need to create a .NET Core CLI tools package instead. Like the NuGet PMC commands, any version of the command should be able to work with any version of the framework.

Known Issues
------------

- [x] System.DllNotFoundException: Unable to load DLL 'sni.dll': The specified module could not be found. (Exception from HRESULT: 0x8007007E) (#4953)
- [x] Unable to load DLL 'sqlite3': The specified module could not be found. (Exception from HRESULT: 0x8007007E) (dotnet/cli#2289)
- [x] Could not load file or assembly 'System.Data.SqlClient, Version=4.1.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies. The system cannot find the file specified. (dotnet/cli#2289)
- [x] Support netcoreapp1.0 apps
- [x] Hosting environment not setup correctly. https://github.com/aspnet/EntityFramework/issues/5065
- [x] Newtonsoft.Json 7.0.0 missing at runtime https://github.com/NuGet/Home/issues/2462
- [x] Newtonsoft.Json 8.0.0 could not be loaded at runtime https://github.com/dotnet/cli/issues/2514. Workaround in #5082 
- [x] ~~Tools using ConfigurationBuilder without setting basepath cannot find config files.~~ won't fix. User's need to set base path to IApplicationEnvironment.ApplicationBasePath.
- [x] Need to change to nested-dispatch model because `AssemblyLoadContext` does not provide the necessary isolation https://github.com/dotnet/cli/issues/2514
- [x] Cannot use dotnet-ef on a netstandard library or standalone app with nested-dispatch https://github.com/dotnet/cli/issues/2645

Usage
-----
Your `project.json` should look something like this.

```JSON
{
    "compilationOptions": {
        "emitEntryPoint": true
    },
    "tools": {
        "Microsoft.EntityFrameworkCore.Tools": { 
            "version": "1.0.0-*",
            "imports": "portable-net452+win81"
         }
    },
    "dependencies": {
        "Microsoft.EntityFrameworkCore.Sqlite": "1.0.0-*",
        "Microsoft.EntityFrameworkCore.Sqlite.Design": { "version": "1.0.0-*", "type": "build" },

        "Microsoft.EntityFrameworkCore.Tools": { "version": "1.0.0-*", "type": "build" }
    },
    "frameworks": {
        "netcoreapp1.0": {
            "imports": "portable-net452+win81",
            "dependencies": {
                "Microsoft.NETCore.App": { 
                     "version": "1.0.0-*",
                     "type": "platform"
                 }
            }
        }
    }
}
```

Your workflow will look something like this.

    dotnet restore
    cd src\MyProject
    dotnet ef migrations add MyMigration
    dotnet ef database update

**Comment by bricelam** [comment]
Open questions: (cc @davidfowl)
* Will it just work with project.json, or *.csproj and raw assemblies (a la #646) too?
* How do we deal with CLR vs CoreCLR and x86 vs x64?

**Comment by AdamDotNet** [comment]
Will this ultimately resolve #3888?

**Comment by bricelam** [comment]
@AdamDotNet No, either way we'll need to author the EF.Commands package (or whatever it ends up being called) to contain a `ref\dotnet` directory.

**Comment by bricelam** [comment]
I'm reopening this. There have been a lot of regressions lately with all the churn. I'll wait until things are consistently stable before closing it .

**Comment by bricelam** [comment]
Known issues:
* Could not load file or assembly...
* dotnet-ef is not compatible with DNXCore

**Comment by capesean** [comment]
Getting error: **No executable found matching command "dotnet-ef"** from build 20051.

**Comment by natemcmaster** [comment]
Btw, I'm getting issues in 1.0.0-rc2-20075 with the wrong sqlite3.dll bitness being loaded.

```
System.BadImageFormatException: An attempt was made to load a program with an incorrect format. (Exception from HRESULT: 0x8007000B)
   at Microsoft.Data.Sqlite.Interop.NativeMethods.Sqlite3_sqlite3.sqlite3_open_v2(IntPtr filename, Sqlite3Handle& ppDb, Int32 flags, IntPtr vfs)
   at Microsoft.Data.Sqlite.Interop.NativeMethods.sqlite3_open_v2(String filename, Sqlite3Handle& ppDb, Int32 flags, String vfs)
   at Microsoft.Data.Sqlite.SqliteConnection.Open()
   at Microsoft.EntityFrameworkCore.Storage.RelationalConnection.Open()
   at Microsoft.EntityFrameworkCore.Storage.Internal.SqliteRelationalConnection.Open()
   at Microsoft.EntityFrameworkCore.Storage.Internal.SqliteDatabaseCreator.Exists()
   at Microsoft.EntityFrameworkCore.Migrations.HistoryRepository.Exists()
   at Microsoft.EntityFrameworkCore.Migrations.HistoryRepository.GetAppliedMigrations()
   at Microsoft.EntityFrameworkCore.Migrations.Design.MigrationsScaffolder.RemoveMigration(String projectDir, String rootNamespace)
   at Microsoft.EntityFrameworkCore.Design.MigrationsOperations.RemoveMigration(String contextType)
   at Microsoft.EntityFrameworkCore.Commands.MigrationsRemoveCommand.<>c__DisplayClass0_0.<Configure>b__0()
   at Microsoft.Extensions.CommandLineUtils.CommandLineApplication.Execute(String[] args)
   at Microsoft.EntityFrameworkCore.Commands.Program.Main(String[] args)
```

**Comment by NTaylorMullen** [comment]
I'm getting the following error when running `dotnet ef database update` on [this](https://github.com/aspnet/benchmarks/tree/nimullen/cli) branch in Benchmarks. Looks like it can't resolve any of the `Startup.cs` required dependencies.

```
C:\Users\nimullen\Documents\GitHub\Benchmarks\src\Benchmarks [nimullen/cli +0 ~1 -0]> dotnet ef database update
System.InvalidOperationException: Unable to resolve service for type 'Microsoft.AspNetCore.Hosting.IHostingEnvironment' while attempting to activate 'Benchmarks.Startup'.
   at Microsoft.Extensions.Internal.ActivatorUtilities.ConstructorMatcher.CreateInstance(IServiceProvider provider)
   at Microsoft.Extensions.Internal.ActivatorUtilities.CreateInstance(IServiceProvider provider, Type instanceType, Object[] parameters)
   at Microsoft.EntityFrameworkCore.Design.Internal.StartupInvoker.Invoke(Type type, String[] methodNames, IServiceCollection services)
   at Microsoft.EntityFrameworkCore.Design.Internal.StartupInvoker.ConfigureServices()
   at Microsoft.EntityFrameworkCore.Design.DbContextOperations..ctor(ILoggerProvider loggerProvider, Assembly assembly, Assembly startupAssembly, String environment)
   at Microsoft.EntityFrameworkCore.Design.MigrationsOperations..ctor(AssemblyLoader assemblyLoader, ILoggerProvider loggerProvider, Assembly assembly, Assembly startupAssembly, String environment, String projectDir, String rootNamespace)
   at Microsoft.EntityFrameworkCore.Commands.OperationExecutor.<>c__DisplayClass3_0.<.ctor>b__4()
   at Microsoft.EntityFrameworkCore.Internal.LazyRef`1.get_Value()
   at Microsoft.EntityFrameworkCore.Commands.OperationExecutor.UpdateDatabase(String targetMigration, String contextType)
   at Microsoft.EntityFrameworkCore.Commands.DatabaseUpdateCommand.<>c__DisplayClass0_0.<Configure>b__0()
   at Microsoft.Extensions.CommandLineUtils.CommandLineApplication.Execute(String[] args)
   at Microsoft.EntityFrameworkCore.Commands.Program.Main(String[] args)
Unable to resolve service for type 'Microsoft.AspNetCore.Hosting.IHostingEnvironment' while attempting to activate 'Benchmarks.Startup'.
```

As a side note, i'm not sure how the tool will ever be smart enough to handle situations were you call `.ConfigureServices` in your `Main`. Do you guys plan to soft-boot the users app?

**Comment by bricelam** [comment]
@NTaylorMullen We've got #4546 to sort out the interaction with `Main` and `Startup`. Our goal is not to work out-of-the-box with 100% of applications. Instead we try and cover the 80% case (ones that follow the patterns established by the templates). Right now, we scan for a class named `Startup` an use that instead of trying to dig into `Main`. As a catch-all, you can take over the design-time instantiation of your `DbContext` by implementing `IDbContextFactory<T>` on any class with a default constructor in the same assembly.

**Comment by bricelam** [comment]
I think I got to the bottom of both the inability to resolve `System.Data.SqlClient.dll` and the missing/incorrect `sqlite3.dll`. We need to specify a fallback list of RIDs when building the `ProjectContext`.

**Comment by NTaylorMullen** [comment]
FYI: I tried the workaround of having an `IDbContextFactory<T>` to prevent construction of my `Startup.cs` but still no go. It attempts to create the `Startup.cs` via the `StartupInvoker` in the [DbContextOperations constructor](https://github.com/aspnet/EntityFramework/blob/62fb0b4704b0970180f0a35e8da3b0b55ad4a5aa/src/Microsoft.EntityFrameworkCore.Commands/Design/DbContextOperations.cs#L44-L45).

Would probably be better to evaluate runtime services via the `Startup.cs` only if there's no available `IDbContextFactory`s available.

**Comment by bricelam** [comment]
@NTaylorMullen agreed. At the very least, we shouldn't terminate if we can't construct `Startup`--just warn and keep looking for `DbContext` classes.

**Comment by bricelam** [comment]
@NTaylorMullen What if you don't specify `--startup-project` and put the `IDbContextFactory` in the data assembly?

**Comment by ahmedanis03** [comment]
dotnet ef database update
sni.dll not found
```
dotnet : System.DllNotFoundException: Unable to load DLL 'sni.dll': The specified module could not be found. (Exception from HRESULT: 0x8007007E)
At line:1 char:1
+ dotnet ef database update
+ ~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (System.DllNotFo...LT: 0x8007007E):String) [], RemoteException
    + FullyQualifiedErrorId : NativeCommandError
 
   at System.Data.SqlClient.SNINativeMethodWrapper.UnmanagedIsTokenRestricted(IntPtr token, Boolean& isRestricted)

   at System.Data.Win32NativeMethods.IsTokenRestrictedWrapper(IntPtr token)

   at System.Data.ProviderBase.DbConnectionPoolIdentity.GetCurrent()

   at System.Data.ProviderBase.DbConnectionPoolGroup.GetConnectionPool(DbConnectionFactory connectionFactory)

   at System.Data.ProviderBase.DbConnectionFactory.GetConnectionPool(DbConnection owningObject, DbConnectionPoolGroup connectionPoolGroup)

   at System.Data.ProviderBase.DbConnectionFactory.TryGetConnection(DbConnection owningConnection, TaskCompletionSource`1 retry, DbConnectionOptions userOptions, DbConnectionInternal oldConnection, DbConnectionInternal& connection)


   at System.Data.ProviderBase.DbConnectionInternal.TryOpenConnectionInternal(DbConnection outerConnection, DbConnectionFactory connectionFactory, TaskCompletionSource`1 retry, DbConnectionOptions userOptions)

   at System.Data.SqlClient.SqlConnection.TryOpen(TaskCompletionSource`1 retry)

   at System.Data.SqlClient.SqlConnection.Open()
   at Microsoft.EntityFrameworkCore.Storage.RelationalConnection.Open()

   at Microsoft.EntityFrameworkCore.Storage.Internal.SqlServerDatabaseCreator.Exists(Boolean retryOnNotExists)

   at Microsoft.EntityFrameworkCore.Migrations.HistoryRepository.Exists()

   at Microsoft.EntityFrameworkCore.Migrations.Internal.Migrator.Migrate(String targetMigration)

   at Microsoft.EntityFrameworkCore.Design.MigrationsOperations.UpdateDatabase(String targetMigration, String contextType)

   at Microsoft.EntityFrameworkCore.Commands.DatabaseUpdateCommand.<>c__DisplayClass0_0.<Configure>b__0()

   at Microsoft.Extensions.CommandLineUtils.CommandLineApplication.Execute(String[] args)

   at Microsoft.EntityFrameworkCore.Commands.Program.Main(String[] args)

Unable to load DLL 'sni.dll': The specified module could not be found. (Exception from HRESULT: 0x8007007E)
```

**Comment by bricelam** [comment]
See also dotnet/cli#1827

**Comment by natemcmaster** [comment]
Might just be the TFM change, but latest cli doesn't even find dotnet-ef. :cry: 
```
No executable found matching command "dotnet-ef"
```
Using 1.0.0-beta-001937

**Comment by bricelam** [comment]
@natemcmaster yeah, I had #4813 out, but it got side-railed by dotnet/cli#1970 which will change again when `netcoreapp` is in and then again a little bit to support `net452`... :anguished:

**Comment by moozzyk** [comment]
looping myself

**Comment by rowanmiller** [comment]
Blocked until the latest CLI bits work in our branches

**Comment by bruceyclee** [comment]
looping myself too

**Comment by mlorbetske** [comment]
Looping myself

**Comment by bricelam** [comment]
We finally took a new build of the CLI! We should be able to make more progress on this issue now.

**Comment by bricelam** [comment]
Things are looking a lot better after #4977, but there are still some dependency resolution issues. :unamused: See the issue description for known issues and usage notes.

**Comment by natemcmaster** [comment]
Blocked on https://github.com/dotnet/cli/issues/2315. The issue: sqlite3.dll and sni.dll are not included in the AssemblyLoadContext, so any calls to actually connect to the database fail.

**Comment by natemcmaster** [comment]
Should be done with changes in #5002

**Comment by natemcmaster** [comment]
Reopening. Users are reporting issues with Newtonsoft.Json and typeload exceptions.

**Comment by moozzyk** [comment]
/cc @victorhurdugaci 
also this https://github.com/aspnet/IISIntegration/issues/133

**Comment by victorhurdugaci** [comment]
Yup, investigating the issue in dotnet watch. For some reason it only fails when you try manually and works fine on the CI

**Comment by victorhurdugaci** [comment]
The dotnet watch issue: https://github.com/aspnet/dotnet-watch/issues/72

**Comment by victorhurdugaci** [comment]
The fix is to add `imports` to the tool and re-restore. Make sure you clear the packages folder first because it doesn't work otherwise, the package has to be re downloaded.

```
"tools": {
        "dotnet-watch": {
           "version": "1.0.0-*",
           "imports": "portable-net451+win8"
        }
    }
```

We believe this will be fixed in a future CLI

**Comment by natemcmaster** [comment]
> We believe this will be fixed in a future CLI

@victorhurdugaci is this because of a known and resolved nuget bug? It seems the lock file for the tool ends up with a bad library entry.
```json
"Newtonsoft.Json/7.0.3": {}
```

**Comment by victorhurdugaci** [comment]
@moozzyk or @pakrym can give more details

**Comment by moozzyk** [comment]
When I talked to Victor I thought it was that we need add an import to netstandard1.5 which would be because we were waiting for this to flow to dotnet: https://github.com/NuGet/NuGet.Client/commit/9573f19ce94c84078e0d0e4f128bb24de7bfe140#diff-8c917f3b4e6bc57dc2667ae784dbb4b3R16
But since it requires imports to "portable-net451+win8" I don't think is this. When I looked at the Newtonsoft.Json package they don't have netstandard1.5/netcoreapp1.0 folder in their libs so I wonder if this can be the issue..

**Comment by natemcmaster** [comment]
Here is the issue: https://github.com/NuGet/Home/issues/2462. dotnet-restore silently fails when missing the right imports. Looks like https://github.com/NuGet/NuGet.Client/pull/468 will make it more obvious dotnet-restore fails to get json.net....but for now, we still need to add "imports" to the tools node.

**Comment by natemcmaster** [comment]
The rest of the Newtonsoft.Json story is here: https://github.com/dotnet/cli/issues/2514

**Comment by cevarief** [comment]
```bash
dotnet ef migrations add Init
```
Output : 
```bash
System.IO.FileNotFoundException: Could not load file or assembly 'Newtonsoft.Json, Version=7.0.0.0, Culture=neutral, PublicKeyToken=30ad4fe6b2a6aeed' or one of its dependencies. The system cannot find the file specified.
File name: 'Newtonsoft.Json, Version=7.0.0.0, Culture=neutral, PublicKeyToken=30ad4fe6b2a6aeed'
...
Could not load file or assembly 'Newtonsoft.Json, Version=7.0.0.0, Culture=neutral, PublicKeyToken=30ad4fe6b2a6aeed' or one of its dependencies. The system cannot find the file specified.
```
When restoring package : ```Package Newtonsoft.Json 8.0.3 is not compatible with netcoreapp1.0 (.NETCoreApp,Version=v1.0).```

Config : 
```json
"frameworks": {
    "netcoreapp1.0": {
      "imports": [
        "dnxcore50",
        "net45"
      ],
      "dependencies": {
        "Microsoft.NETCore.App": {
          "type": "platform",
          "version": "1.0.0-*"
        }
      }
    }
  },
  "tools": {
    "dotnet-ef": "1.0.0-*"
  },
```

**Comment by cevarief** [comment]
Is there workaround to make it works ?

**Comment by cevarief** [comment]
I cleared all the nuget package and restore. The newton.json error has gone.
Another error when adding migration : 
```bash 
System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation. ---> System.ArgumentNullException: Value cannot be null.
Parameter name: assemblyName
   at System.Reflection.AssemblyName..ctor(String assemblyName)
   at Microsoft.AspNetCore.Mvc.Internal.DefaultAssemblyPartDiscoveryProvider.DiscoverAssemblyParts(String entryPointAssemblyName)
```

**Comment by patrickmichalina** [comment]
@cevarief I was able to get the tools started using this:

```
"tools": {
    "dotnet-ef": {
      "version":"1.0.0-*",
      "imports":["portable-net40+sl5+win8+wp8+wpa81"]
    }
  }
```

**Comment by patrickmichalina** [comment]
Nevermind... That just throws this...
![capture](https://cloud.githubusercontent.com/assets/6701211/14766424/0c92a96a-09cf-11e6-9364-e6eaea6e2a8a.PNG)

**Comment by moozzyk** [comment]
You used an incorrect import `"portable-net40+sl5+win8+wp8+wpa81"` brought the Silverlight version. You need the import that has "dnxcore50" in it. Note that after you fix your import you have to clean your NuGet cache otherwise the error will not go away (you can clean all NuGet caches using `nuget.exe locals -Clear all`)

**Comment by cevarief** [comment]
I use the following import, but still can not successfully add migration. ```System.ArgumentNullException: Value cannot be null. Parameter name: assemblyName```.
```json
"dotnet-ef": {
      "version": "1.0.0-*",
      "imports": [
        "portable-net45+wp80+win8+wpa81+dnxcore50",
        "portable-net45+win8+wp8+wpa81",
        "portable-net45+win8+wp8"
      ]
    },
```

**Comment by cevarief** [comment]
@mightea Thanks. It works by commenting out addMvc() and addCors().
 ```bash
dotnet ef migration add Init
Done. To undo this action, use 'ef migrations remove'```

**Comment by natemcmaster** [comment]
Heads up to anyone following this issue. We recently made big changes in #5142.  "dotnet-ef" has been renamed to "Microsoft.EntityFrameworkCore.Tools". I've updated the sample usage in the issue description above to show what needs to be in project.json to use the tool now. https://github.com/aspnet/EntityFramework/issues/3925#issue-119549607

I'm leaving this issue open as we are still waiting on partners for fixes so that we can validate and ensure EF tools work.

**Comment by cevarief** [comment]
@natemcmaster What feed should we use to resolve Microsoft.EntityFrameworkCore.Tools, i use both aspnetcirelease and aspnetcidev and get this error : 
```
Unable to resolve 'Microsoft.EntityFrameworkCore.Tools (>= 1.0.0)' for '.NETCoreApp,Version=v1.0'.
Unable to resolve 'Microsoft.EntityFrameworkCore.Tools (>= 1.0.0)' for '.NETCoreApp,Version=v1.0'.
```

**Comment by natemcmaster** [comment]
It has to move through our build system. The fastest I've seen it make it through to CI feeds is 6 hours. I pushed the code about 90 min ago

**Comment by cevarief** [comment]
I've just tried to restore and now it works good. Thanks a lot 👍

**Comment by natemcmaster** [comment]
Closing as we have finished with the implementation of .NET Core CLI commands. Any remaining bugs should be opened as a separate issue.

Additional work for post-RC2:
https://github.com/aspnet/EntityFramework/issues/5188
https://github.com/aspnet/EntityFramework/issues/5187
https://github.com/aspnet/EntityFramework/issues/5190

**Comment by gtbull80** [comment]
not sure if this is the right place for this...but any ideas?

PM> dotnet ef migrations add MyFirstMigration
dotnet.exe : Build failed.
At line:1 char:7
+ dotnet <<<<  ef migrations add MyFirstMigration
    + CategoryInfo          : NotSpecified: (Build failed.:String) [], RemoteException
    + FullyQualifiedErrorId : NativeCommandError

**Comment by natemcmaster** [comment]
@gtbull80 dotnet-ef requires your project to build before it can add a migration. Run `dotnet build` to see the compilation errors.

**Comment by gtbull80** [comment]
Yeah...So I'm trying to wrap my head around all the new tech.  Life's fun on the bleeding edge!  When I run `dotnet build` , there are no errors...in fact there's no output at all.  So I'm not sure if my environment configuration is wrong or I'm missing something else.  I run the command in VS studio's package manager console from my project src\\{projectname} directory.  I'm running an MVC6 app targeting RC2.  My project.json:

```
{
    "userSecretsId": "aspnet5-EntityFrameworkTest-cd16c4bc-1228-4400-8b20-3c4076e2ed42",
    "version": "1.0.0-*",
    "compilationOptions": {
        "emitEntryPoint": true
    },

    "dependencies": {
        "AspNet.Security.OAuth.Validation": "1.0.0-*",
        "IdentityModel": "2.0.0-*",
        "Microsoft.AspNetCore.Authentication.Cookies": "1.0.0-*",
        "Microsoft.AspNetCore.Authentication.JwtBearer": "1.0.0-*",
        "Microsoft.AspNetCore.Cors": "1.0.0-*",
        "Microsoft.AspNetCore.Diagnostics": "1.0.0-*",
        "Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore": "1.0.0-*",
        "Microsoft.AspNetCore.Identity.EntityFrameworkCore": "1.0.0-*",
        "Microsoft.AspNetCore.IISPlatformHandler": "1.0.0-*",
        "Microsoft.AspNetCore.Mvc": " 1.0.0-*",
        "Microsoft.AspNetCore.Mvc.Dnx": "1.0.0-*",
        "Microsoft.AspNetCore.Mvc.TagHelpers": "1.0.0-*",
        "Microsoft.AspNetCore.Razor": "1.0.0-*",
        "Microsoft.AspNetCore.Server.Kestrel": "1.0.0-*",
        "Microsoft.AspNetCore.StaticFiles": "1.0.0-*",
        "Microsoft.EntityFrameworkCore": "1.0.0-*",
        "Microsoft.EntityFrameworkCore.Commands": "1.0.0-rc3-20680",
        "Microsoft.EntityFrameworkCore.SqlServer": "1.0.0-*",
        "Microsoft.Extensions.CodeGenerators.Mvc": "1.0.0-*",
        "Microsoft.Extensions.Configuration.FileProviderExtensions": "1.0.0-*",
        "Microsoft.Extensions.Configuration.Json": "1.0.0-*",
        "Microsoft.Extensions.Configuration.UserSecrets": "1.0.0-*",
        "Microsoft.Extensions.Logging": "1.0.0-*",
        "Microsoft.Extensions.Logging.Console": "1.0.0-*",
        "Microsoft.Extensions.Logging.Debug": "1.0.0-*",
        "Microsoft.VisualStudio.Web.BrowserLink.Loader": "14.0.0-*",
        "OpenIddict.Core": "1.0.0-*",
        "OpenIddict.EF": "1.0.0-*",
        "System.IdentityModel.Tokens": "5.0.0-*"       
    },

    "commands": {
        "web": "OpenIddictAuthorizationServerTest",
        "ef": "Microsoft.EntityFrameworkCore.Commands"
    },

    "frameworks": {
        "netcoreapp1.0": {
            "imports": "portable-net452+win81",
            "dependencies": {
                "Microsoft.NETCore.App": {
                    "version": "1.0.0-*"
                }
            }
        }
    },

    "exclude": [
        "wwwroot",
        "node_modules"
    ],
    "publishExclude": [
        "**.user",
        "**.vspscc"
    ],
    "scripts": {
        "prepublish": [ "npm install", "bower install", "gulp clean", "gulp min" ]
    },
    "tools": {
        "dotnet-ef": "1.0.0-*"
    }
}
```
